### PR TITLE
add jax core test

### PR DESCRIPTION
**Pull request by kiran-thumma**
## Motivation
- Add an automated Core GPU validation test that verifies built JAX wheel (.whl) files are functional.
- We currently lack automated wheel-level GPU validation. This test will catch regressions in packaging, compatibility issues, missing binaries, or other wheel-level failures before wheels are promoted.
- Faster feedback, reduces risk of shipping broken GPU wheels.

## Technical Details

- New workflow: test_linux_jax_wheels.yml
- Test type: Core tests
- Purpose: Install build JAX wheels in a docker and run tests.
- Diagram:
   jax-test-workflow-11-06 024529

## Test Plan

- Local Verification
- CI Verification

## Test Result

- Workflow run: 
- ENV Tested: 
- Output Summary:

## Submission Checklist

- [x] Core tests pass on at least one GPU configuration in CI
- [x] Workflow has a manual dispatch input for testing arbitrary wheel URLs

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

**Comment by kiran-thumma** [comment]
Latest Run : https://github.com/ROCm/TheRock/actions/runs/19225042724

File: .github/workflows/test_linux_jax_wheels.yml [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        working-directory: jax
+        env:
+          ROCM_VERSION: ${{ inputs.rocm_version }}
+        run: |
+            # Run the tests and capture output
+            CORETEST_OUTPUT=$(python3 build/ci_build test \
+            "${JAX_IMAGE}" \
+            --test-cmd "pytest jax_tests/tests/core_test.py" 2>&1)
+            echo "$CORETEST_OUTPUT" > coretest_output.txt
+            cat coretest_output.txt >> $GITHUB_STEP_SUMMARY
```
Comment (ScottTodd) [inline .github/workflows/test_linux_jax_wheels.yml]: This step doesn't make sense to me. Why capture the output just to dump it out to `GITHUB_STEP_SUMMARY`?

How about this, streaming output to stdout as is the default?

```suggestion
      - name: Run unit tests
        working-directory: jax
        env:
          ROCM_VERSION: ${{ inputs.rocm_version }}
        run: |
            python3 build/ci_build test \
              "${JAX_IMAGE}" \
              --test-cmd "pytest jax_tests/tests/core_test.py"
```

You could _also_ output to a file and then upload that file, as we do with https://github.com/ROCm/TheRock/blob/main/build_tools/teatime.py and https://github.com/ROCm/TheRock/blob/main/build_tools/github_actions/post_build_upload.py for our CMake builds.

Only capturing the output in a file and then posting makes the logs on the test step itself completely empty: https://github.com/ROCm/TheRock/actions/runs/19225042724/job/54954865305#step:9:1. For long running test processes being able to see intermediate output as it appears is very useful, and we shouldn't have just one workflow (test_linux_jax_wheels.yml) handling logging differently.

File: .github/workflows/test_linux_jax_wheels.yml [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+      - name: Prepare wheelhouse and download JAX wheels
+        shell: bash
+        working-directory: jax
+        run: |
+          python3 ../build_tools/fetch_wheels.py \
+            --cloudfront_url ${{ inputs.package_index_url }} \
+            --amdgpu_family ${{ inputs.amdgpu_family }} \
+            --dir "wheelhouse" \
+            --list_whls '${{ inputs.jax_whl_list }}'
```
Comment (ScottTodd) [inline .github/workflows/test_linux_jax_wheels.yml:125]: What is this actually doing with the downloaded wheels? Is the `python3 build/ci_build` script using this "wheelhouse" directory?

Can this use [`setup_venv.py`](https://github.com/ROCm/TheRock/blob/main/build_tools/setup_venv.py) as we do already in the pytorch test workflow?

https://github.com/ROCm/TheRock/blob/2e9e16862c3f0f250631b19e50166a031f7f5f07/.github/workflows/test_pytorch_wheels.yml#L112-L118

Here are some logs of that script running in the workflow: https://github.com/ROCm/TheRock/actions/runs/19228713043/job/54971284725#step:7:22

File: build_tools/fetch_wheels.py
Comment (ScottTodd) [inline build_tools/fetch_wheels.py:1]: I don't think this script should be needed, see my other comment.

If we do keep this script, some high level comments:

* For downloading files, I'd suggest using either `pip download` (https://pip.pypa.io/en/stable/cli/pip_download/) or `requests` (https://pypi.org/project/requests/), instead of `subprocess.run` with `wget` (we also have helper `exec` functions that set better defaults for `subprocess`)
* Use `pathlib` instead of `os` and raw strings, consistently throughout

File: .github/workflows/build_linux_jax_wheels.yml [hunk 3]
```diff
@@ -117,6 +148,13 @@ jobs:
           aws-region: us-east-2
           role-to-assume: arn:aws:iam::692859939525:role/therock-${{ inputs.release_type }}-releases
 
+      - name: jax wheel list
+        id: jax_wheel_list
+        run: |
+          export jax_whl_list="$(python3 ./build_tools/mapping_built_jax_wheels.py --dir "${{ env.PACKAGE_DIST_DIR }}")"
+          echo "jax_whl_list=${jax_whl_list}" >> "$GITHUB_OUTPUT"
+          echo "${jax_whl_list}" >> "$GITHUB_STEP_SUMMARY"
```
Comment (ScottTodd) [inline .github/workflows/build_linux_jax_wheels.yml:156]: Please follow the style of https://github.com/ROCm/TheRock/blob/main/build_tools/github_actions/write_torch_versions.py here with a `build_tools/github_actions/write_jax_versions.py`

Usage:
* https://github.com/ROCm/TheRock/blob/2e9e16862c3f0f250631b19e50166a031f7f5f07/.github/workflows/build_portable_linux_pytorch_wheels.yml#L177-L188
* https://github.com/ROCm/TheRock/blob/2e9e16862c3f0f250631b19e50166a031f7f5f07/.github/workflows/build_portable_linux_pytorch_wheels.yml#L117-L122
* (and other parts of that file)
* https://github.com/ROCm/TheRock/blob/2e9e16862c3f0f250631b19e50166a031f7f5f07/.github/workflows/test_pytorch_wheels.yml#L80-L82
* https://github.com/ROCm/TheRock/blob/2e9e16862c3f0f250631b19e50166a031f7f5f07/.github/workflows/test_pytorch_wheels.yml#L112-L118

Note that the _versions_ of the files are what is computed and passed through the workflows, not their filenames. Then `pip` commands use the versions to download or install wheels with those versions (as users would do outside of github actions workflows)

File: .github/workflows/test_linux_jax_wheels.yml [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      - name: Compute ROCM_VERSION_SHORT
+        working-directory: jax
+        env:
+            ROCM_VERSION: ${{ inputs.rocm_version }}
+        run: |
+            # Extract major.minor.patch from ROCM_VERSION
+            ROCM_VERSION_SHORT="$(echo "${ROCM_VERSION}" | sed -E 's/^([0-9]+)\.([0-9]+)\.([0-9]+).*/\1.\2.\3/')"
+            echo "ROCM_VERSION_SHORT=${ROCM_VERSION_SHORT}" >> "$GITHUB_ENV"
+            echo "Full ROCm version: ${ROCM_VERSION}"
+            echo "Using semantic ROCm version: ${ROCM_VERSION_SHORT}"
```
Comment (ScottTodd) [inline .github/workflows/test_linux_jax_wheels.yml:136]: This code is suspicious and I don't think it should be needed. At a minimum there should be a comment here explaining what problem this code is solving.

File: .github/workflows/test_linux_jax_wheels.yml [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        run: |
+            python3 build/ci_build \
+            --rocm-version="${ROCM_VERSION_SHORT}" \
+            --therock-path="${{ inputs.tar_url }}" \
+            build_dockers \
+            -f ubu24
+            # Assign JAX_IMAGE to the expected image name produced by the build script
+            JAX_IMAGE="jax-ubu24.rocm$(echo "${ROCM_VERSION_SHORT}" | tr -d '.'):latest"
+            echo "JAX_IMAGE=${JAX_IMAGE}" >> "$GITHUB_ENV"
+            echo "Built JAX Docker image: ${JAX_IMAGE}"
```
Comment (ScottTodd) [inline .github/workflows/test_linux_jax_wheels.yml:151]: Why is a Docker image needed here?

If keeping the docker image, a tag other than `:latest` is probably called for (I assume this doesn't get uploaded anywhere but still safer to not have a test workflow directly set the tag to "latest")

File: .github/workflows/test_linux_jax_wheels.yml (hunk 2 above)
Comment (kiran-thumma) [reply .github/workflows/test_linux_jax_wheels.yml:125]: No, we cannot use a setup_venv.py. As we have to change the docker build steps all that if we have to use a venv. This is the same way just a wheelhouse directory.

File: build_tools/fetch_wheels.py
Comment (kiran-thumma) [reply build_tools/fetch_wheels.py:1]: We will need this script and will update as per your comments.

File: .github/workflows/test_linux_jax_wheels.yml (hunk 5 above)
Comment (kiran-thumma) [reply .github/workflows/test_linux_jax_wheels.yml:151]: The core_tests and other GPU tests are ran using the docker image. Even if we test the docker image to latest or not the latest, it should be any difference as its running inside a pod and will be decommissioned after this workflow is completed running.

File: .github/workflows/test_linux_jax_wheels.yml (hunk 2 above)
Comment (ScottTodd) [reply .github/workflows/test_linux_jax_wheels.yml:125]: Why is Docker a requirement? I still haven't seen a clear answer to that. What are the user install instructions like? Are they more than `pip install jax --index-url=...`? Tests should match what users will do, and we can't expect users to download files manually then build a dockerfile.

File: .github/workflows/build_linux_jax_wheels.yml (hunk 3 above)
Comment (kiran-thumma) [reply .github/workflows/build_linux_jax_wheels.yml:156]: I will think of adding a logging or summary later as part of enhancements as it already has the logs.

**Comment by kiran-thumma** [comment]
Latest Run : https://github.com/ROCm/TheRock/actions/runs/19281849915

File: .github/workflows/test_linux_jax_wheels.yml (hunk 2 above)
Comment (kiran-thumma) [reply .github/workflows/test_linux_jax_wheels.yml:125]: Yes, it does has other requriements to be configured and not the whole code is in theROCk as we are using this workflow as a wrapper to the rocm-jax and upstream code. I have update the script as per suggestion to use requests instead of wget.

File: .github/workflows/test_linux_jax_wheels.yml (hunk 4 above)
Comment (kiran-thumma) [reply .github/workflows/test_linux_jax_wheels.yml:136]: If I'm passing the full rocmversion docker fails to add that as a tag. Added more comments.

File: .github/workflows/test_linux_jax_wheels.yml (hunk 5 above)
Comment (marbre) [reply .github/workflows/test_linux_jax_wheels.yml:151]: Why does this need to use a custom Docker image? I strongly suggest to follow the style used in testing torch wheels in TheRock. There a Docker image is only used to guarantee having a ROCm-free environment. Packages to test should than be installed via pip. See https://github.com/ROCm/TheRock/blob/main/.github/workflows/test_pytorch_wheels.yml

File: .github/workflows/test_linux_jax_wheels.yml (hunk 4 above)
Comment (marbre) [reply .github/workflows/test_linux_jax_wheels.yml:136]: There is no need to compute a tag. Rather spin up a docker, pip install the jax wheels, run tests and that's it. No need to _build_ a docker container.

**Review by marbre — COMMENTED** [review]
I think we're hitting the same issue as with the previous PR. This has a diagram but doesn't argue why it needs to use Docker the way it does. I really think this needs to follow the workflow style used for testing torch wheels.

File: .github/workflows/test_linux_jax_wheels.yml (hunk 4 above)
Comment (kiran-thumma) [reply .github/workflows/test_linux_jax_wheels.yml:136]: Either way its the same. Goal is to test the wheels. The tests currently which jax team is supporting are these tests.

File: .github/workflows/test_linux_jax_wheels.yml (hunk 5 above)
Comment (kiran-thumma) [reply .github/workflows/test_linux_jax_wheels.yml:151]: This is not a custom docker image. Its the same docker file which jax upstream supports and the tests which we are running. The tests currently are running on a docker image based. I can have a new issue which is to enhance it the same way as torch. Most of the tests support only docker image based even the performance and accuracy tests. for JAX-MAX training. Even for the pytorch we have to build the docker image to support performance training pytorch which is PRIMUS.

**Comment by kiran-thumma** [comment]
> I think we're hitting the same issue as with the previous PR. This has a diagram but doesn't argue why it needs to use Docker the way it does. I really think this needs to follow the workflow style used for testing torch wheels.

I have explained the same thing as the previous PR. To run the tests which currently JAX team supports are based on a docker image. And the we have to use the same dockerfile which is from upstream of JAX. Thats the reason I have added a flow diagram. Please check the flow diagram. Thats the flow if you need changes we have to check with the upstream jax and jax team.

File: .github/workflows/test_linux_jax_wheels.yml (hunk 4 above)
Comment (ScottTodd) [reply .github/workflows/test_linux_jax_wheels.yml:136]: Looks like JAX has these docs, which do suggest running in a container: https://github.com/ROCm/rocm-jax/blob/master/BUILDING.md#3-running-tests, but I'm not seeing a technical explanation there for why tests are difficult to run outside of a container. We are strongly discouraging Docker and container usage in new packaging in TheRock and this is the time to investigate those details.

Again, and I'm getting frustrated repeating myself - this type of work needs to **start with a design discussion, before we get into details during code review**. We've wasted multiple weeks now going back and forth without seeing a plan written down. At a minimum provide links to pages like these so we can discuss the design points - that is your job as the PR author and not something reviewers should need to research on their own:
* https://rocm.docs.amd.com/projects/install-on-linux/en/latest/install/3rd-party/jax-install.html (in contrast to https://rocm.docs.amd.com/projects/install-on-linux/en/latest/install/3rd-party/pytorch-install.html#test-the-pytorch-installation)
* https://github.com/ROCm/rocm-jax/blob/master/BUILDING.md#3-running-tests

File: .github/workflows/test_linux_jax_wheels.yml (hunk 2 above)
Comment (charleshofer) [reply .github/workflows/test_linux_jax_wheels.yml:125]: @ScottTodd It's not a requirement for running tests. This is just the way that we currently do it in our CI because some Ubuntu docker images are part of our deliverables. We do everything through the `build/ci_build` script, and the script assumes that you want to do the build in three stages:

1. Build the wheels
2. Build the docker images
3. Test using the docker images

There isn't a step in our CI that installs the wheels (and requirements like ROCm or pytest) with pip and then runs the tests. You can do this, of course, but we just don't have an easy command for it in `build/ci_build` because we're taking care of that in step 3 of our CI. We could probably add a command that will do that to `build/ci_build` though, or you could try and do that right in the workflow.

File: .github/workflows/test_linux_jax_wheels.yml (hunk 4 above)
Comment (charleshofer) [reply .github/workflows/test_linux_jax_wheels.yml:136]: They're not terribly difficult, it just requires the right dependencies. Building the image and then running with that is just what we do in our current CI setup. See my other comment. I wholeheartedly agree that doing some design work and getting clear on how TheRock expects frameworks to build and test is much needed, and I'd be happy to discuss it somewhere that's not a PR comment thread.

**Comment by kiran-thumma** [comment]
Latest Run : https://github.com/ROCm/TheRock/actions/runs/19287934450

As per our Discussion created a new issue for enhancement which will be by next release. https://github.com/ROCm/TheRock/issues/2124

**Comment by kiran-thumma** [comment]
Closing this PR as its been reworked and as suggested started a new PR with changes https://github.com/ROCm/TheRock/pull/2247
